modules/core/src/main/scala/com.snowplowanalytics.iglu/client/validator/CirceValidator.scala

```diff
@@ -240,4 +247,71 @@ object CirceValidator extends Validator[Json] {
 
   private def fromValidationMessage(m: ValidationMessage): ValidatorReport =
     ValidatorReport(m.getMessage, m.getPath.some, m.getArguments.toList, m.getType.some)
+
+  type InitValidatorCache[F[_]] =
+    CreateLruMap[F, (SchemaKey, Int), Either[ValidatorError.InvalidSchema, JsonSchema]]
+  def validatorF[F[_]: Monad: InitValidatorCache]: F[ValidatorF[F, Json]] =
+    CreateLruMap[F, (SchemaKey, Int), Either[ValidatorError.InvalidSchema, JsonSchema]]
+      .create(100)
+      .map(validatorFImpl(_))
+
+  def validatorFImpl[F[_]: Monad](
+    lookupJsonSchema: LruMap[F, (SchemaKey, Int), Either[ValidatorError.InvalidSchema, JsonSchema]]
+  ): ValidatorF[F, Json] =
+    new ValidatorF[F, Json] {
+
+      def lookupOrEvaluate(
```

Comment:
What do you think about sth like that:

```scala
      private def lookupOrEvaluate(result: ResolverResult[Json]): F[Either[ValidatorError, JsonSchema]] = {
        result match {
          case Cached(key, schema, timestamp) =>
            evaluationCache.get((key, timestamp)).flatMap {
              case Some(alreadyEvaluatedSchema) => 
                alreadyEvaluatedSchema.pure[F]
              case None =>  
               evaluateSchema(schema)
                 .flatTap(result => evaluationCache.put((key, timestamp), result))
            }
          case NotCached(schema) =>
            evaluateSchema(schema)
        }
      }

  private def evaluateSchema[F[_] : Monad](schema: Json) = {
    Either
      .catchNonFatal(
        IgluMetaschemaFactory
          .getSchema(circeToJackson(schema), SchemaValidatorsConfig)
      )
      .leftMap(ValidatorError.schemaIssue)
      .pure[F]
  }

```
?

Reply:
This is a definite improvement to my version!